test/e2e/persistent_volumes.go

```diff
@@ -62,35 +77,41 @@ func nfsServerPodCleanup(c clientset.Interface, config VolumeTestConfig) {
 	}
 }
 
-// Cleanup up pvs and pvcs in multi-pv-pvc test cases. All entries found in the pv and
+// Clean up a pv and pvc in a single pv/pvc test case.
+func pvPvcCleanup(c clientset.Interface, ns string, pv *api.PersistentVolume, pvc *api.PersistentVolumeClaim) {
+	deletePersistentVolumeClaim(c, pvc.Name, ns)
+	deletePersistentVolume(c, pv.Name)
+}
+
+// Clean up pvs and pvcs in multi-pv-pvc test cases. All entries found in the pv and
 // claims maps are deleted.
-// Note: this is the only code that deletes PV objects.
-func pvPvcCleanup(c clientset.Interface, ns string, pvols pvmap, claims pvcmap) {
+func pvPvcMapCleanup(c clientset.Interface, ns string, pvols pvmap, claims pvcmap) {
+	for pvcKey := range claims {
+		deletePersistentVolumeClaim(c, pvcKey.Name, ns)
+		delete(claims, pvcKey)
+	}
 
-	if c != nil && len(ns) > 0 {
-		for pvcKey := range claims {
-			_, err := c.Core().PersistentVolumeClaims(pvcKey.Namespace).Get(pvcKey.Name)
-			if !apierrs.IsNotFound(err) {
-				Expect(err).NotTo(HaveOccurred())
-				framework.Logf("   deleting PVC %v ...", pvcKey)
-				err = c.Core().PersistentVolumeClaims(pvcKey.Namespace).Delete(pvcKey.Name, nil)
-				Expect(err).NotTo(HaveOccurred())
-				framework.Logf("   deleted  PVC %v", pvcKey)
-			}
-			delete(claims, pvcKey)
-		}
+	for pvKey := range pvols {
+		deletePersistentVolume(c, pvKey)
+		delete(pvols, pvKey)
+	}
+}
 
-		for name := range pvols {
-			_, err := c.Core().PersistentVolumes().Get(name)
-			if !apierrs.IsNotFound(err) {
-				Expect(err).NotTo(HaveOccurred())
-				framework.Logf("   deleting PV %v ...", name)
-				err = c.Core().PersistentVolumes().Delete(name, nil)
-				Expect(err).NotTo(HaveOccurred())
-				framework.Logf("   deleted  PV %v", name)
-			}
-			delete(pvols, name)
-		}
+// Delete the PV.
+func deletePersistentVolume(c clientset.Interface, pvName string) {
+	if c != nil && len(pvName) > 0 {
+		framework.Logf("Deleting PersistentVolume %v", pvName)
+		err := c.Core().PersistentVolumes().Delete(pvName, nil)
+		Expect(apierrs.IsNotFound(err)).To(BeTrue())
+	}
+}
+
+// Delete the Claim
+func deletePersistentVolumeClaim(c clientset.Interface, pvcName string, ns string) {
+	if c != nil && len(pvcName) > 0 {
+		framework.Logf("Deleting PersistentVolumeClaim %v", pvcName)
+		err := c.Core().PersistentVolumeClaims(ns).Delete(pvcName, nil)
+		Expect(apierrs.IsNotFound(err)).To(BeTrue())
```

Comment:
This is too short :-). Now you expect that `Delete()` **always** returns `IsNotFound` error, which is not the case. In addition, failure here does not log the failure:

```
I1117 13:20:47.643]   [k8s.io] PersistentVolumes:GCEPD
I1117 13:20:47.644] go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:826[0m
I1117 13:20:47.644]  should test that deleting the PV before the pod does not cause pod deletion to fail on PD detach [It]
I1117 13:20:47.644]     go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/persistent_volumes.go:731
I1117 13:20:47.644] 
I1117 13:20:47.644]    Expected
I1117 13:20:47.644]         <bool>: false
I1117 13:20:47.644]     to be true
I1117 13:20:47.644] 
```

`Expected  <bool>: false to be true` does not show what went wrong. This will do:

```
if err != nil && !apierrs.IsNotFound(err) {
   // log the error & fail
   Expect(err).NotTo(HaveOccurred())
}
```

This applies also to all `IsNotFound` checks below.